spec/JSCompilerSpec.hs

```diff
@@ -0,0 +1,173 @@
+            "throw new MuPatternMatchError() " ++
+          "}catch($error) { if($error.constructor === MuReturn) { return $error.value } else { throw $error } } }"
+          )
+        (Function "f" [Equation [VariablePattern "x", VariablePattern "y"] (UnguardedBody $ MuNumber 5)]) `shouldBeCompiledTo` (
+          "function f() {try { " ++
+            "if(arguments.length === 2 && function(){ return true }(arguments[0]) && function(){ return true }(arguments[1])){ var x = arguments[0];var y = arguments[1]; return new MuNumber(5.0) } " ++
+            "throw new MuPatternMatchError() " ++
+          "}catch($error) { if($error.constructor === MuReturn) { return $error.value } else { throw $error } } }"
+          )
+        (Function "f" [Equation [] (UnguardedBody (Sequence [If (Reference "c") (Return (Reference "x")) (Sequence []), Reference "y"]))]) `shouldBeCompiledTo` (
+          "function f() {try { " ++
+            "if(arguments.length === 0){  " ++
+              "return function(){ " ++
+                "function(){ if(c) { return function(){ throw new MuReturn(x) }() } else { return undefined } }(); " ++
+                "return y " ++
+              "}() " ++
+            "} " ++
+            "throw new MuPatternMatchError() " ++
+          "}catch($error) { if($error.constructor === MuReturn) { return $error.value } else { throw $error } } }"
+          )
+        pending --TODO: Other cases
+
+      it "Procedures" $ do pending
+
+      it "Variable" $ do
+        (Variable "x" $ Reference "y") `shouldBeCompiledTo` "var x = y"
+
+      it "Assignment" $ do
+        (Assignment "x" $ Reference "y") `shouldBeCompiledTo` "x = y"
+
+      it "Reference" $ do
+        (Reference "x") `shouldBeCompiledTo` "x"
+
+      it "Application" $ do
+        (Application (Reference "f") []) `shouldBeCompiledTo` "f()"
+        (Application (Reference "f") [Reference "x", Reference "y"]) `shouldBeCompiledTo` "f(x, y)"
+        (Application (Application (Reference "f") [Reference "x"]) [Application (Reference "g") [Reference "y"]]) `shouldBeCompiledTo` "f(x)(g(y))"
+
+      it "Send" $ do
+        (Send (Reference "o") (Reference "m") []) `shouldBeCompiledTo` "o['m']()"
+        (Send (Reference "o") (Reference "m") [Reference "x", Reference "y"]) `shouldBeCompiledTo` "o['m'](x, y)"
+        (Send (Send (Reference "o") (Reference "m") [Reference "x"]) (Reference "n") [Reference "y"]) `shouldBeCompiledTo` "o['m'](x)['n'](y)"
+
+      it "New" $ do
+        (New "C" []) `shouldBeCompiledTo` "new C()"
+        (New "C" [Reference "x", Reference "y"]) `shouldBeCompiledTo` "new C(x, y)"
+
+      it "Raise" $ do
+        (Raise $ New "E" []) `shouldBeCompiledTo` "function(){ throw new E() }()"
+
+      it "Print" $ do
+        (Print $ Reference "x") `shouldBeCompiledTo` "console.log(x)"
+
+      it "MuNull" $ do
+        MuNull `shouldBeCompiledTo` "new MuNull()"
+
+      it "Sequence" $ do pending
```

Comment:
Sequences are being compiled, the tests are the ones pending.

I did pretty much that, but prepended a return to the last element so it would leave the context with that value, because I wasn't sure if the parsers where inserting Return statements on languages with default return.